internal/codespaces/codespaces.go

```diff
@@ -45,26 +45,30 @@ func ConnectToLiveshare(ctx context.Context, progress progressIndicator, session
 			return nil, fmt.Errorf("error starting codespace: %w", err)
 		}
 	}
-	expBackoff := backoff.NewExponentialBackOff()
 
+	expBackoff := backoff.NewExponentialBackOff()
 	expBackoff.Multiplier = 1.1
 	expBackoff.MaxInterval = 10 * time.Second
 	expBackoff.MaxElapsedTime = 5 * time.Minute
+	ctxBackoff := backoff.WithContext(expBackoff, ctx)
 
-	for retries := 0; !connectionReady(codespace); retries++ {
-		if retries > 1 {
-			duration := expBackoff.NextBackOff()
-			time.Sleep(duration)
-		}
-
-		if expBackoff.GetElapsedTime() >= expBackoff.MaxElapsedTime {
-			return nil, errors.New("timed out while waiting for the codespace to start")
+	err = backoff.Retry(func() error {
+		if connectionReady(codespace) {
+			return nil
 		}
-
 		codespace, err = apiClient.GetCodespace(ctx, codespace.Name, true)
 		if err != nil {
-			return nil, fmt.Errorf("error getting codespace: %w", err)
+			return backoff.Permanent(fmt.Errorf("error getting codespace: %w", err))
+		}
+		return errors.New("codespace not ready yet")
+	}, ctxBackoff)
```

Comment:
If I read this right, I think this could result in a regression in this polling duration

That's because we fetch the codespace on 59, but even if it is ready after re-fetching we will still force it through the retry once more and pause for the retry delay.

Reply:
That's a good point. Will address that!